core/slim_sim.cpp

```diff
@@ -6191,13 +6202,106 @@ void SLiMSim::WritePopulationTable(tsk_table_collection_t *p_tables)
 	}
 }
 
+void SLiMSim::WriteTreeSequenceMetadata(tsk_table_collection_t *p_tables)
+{
+    int ret = 0;
+
+    //////
+    // Top-level (tree sequence) metadata:
+    // In the future, we might need to *add* to the metadata *and also* the schema,
+    // leaving other keys that might already be there.
+    // But that's being a headache, so we're skipping it.
+    nlohmann::json metadata;
+	metadata["SLiM"]["model_type"] = (ModelType() == SLiMModelType::kModelTypeWF) ? "WF" : "nonWF";
+	metadata["SLiM"]["generation"] = Generation();
+	metadata["SLiM"]["file_version"] = SLIM_TREES_FILE_VERSION;
+    if (spatial_dimensionality_ == 0) {
+        metadata["SLiM"]["spatial_dimensionality"] = "";
+    } else if (spatial_dimensionality_ == 1) {
+        metadata["SLiM"]["spatial_dimensionality"] = "x";
+    } else if (spatial_dimensionality_ == 2) {
+        metadata["SLiM"]["spatial_dimensionality"] = "xy";
+    } else {
+        metadata["SLiM"]["spatial_dimensionality"] = "xyz";
+    }
+    if (periodic_x_ & periodic_y_ & periodic_z_) {
+        metadata["SLiM"]["spatial_periodicity"] = "xyz";
+    } else if (periodic_x_ & periodic_y_) {
+        metadata["SLiM"]["spatial_periodicity"] = "xy";
+    } else if (periodic_x_ & periodic_z_) {
+        metadata["SLiM"]["spatial_periodicity"] = "xz";
+    } else if (periodic_y_ & periodic_z_) {
+        metadata["SLiM"]["spatial_periodicity"] = "yz";
+    } else if (periodic_x_) {
+        metadata["SLiM"]["spatial_periodicity"] = "x";
+    } else if (periodic_y_) {
+        metadata["SLiM"]["spatial_periodicity"] = "y";
+    } else if (periodic_z_) {
+        metadata["SLiM"]["spatial_periodicity"] = "z";
+    } else {
+        metadata["SLiM"]["spatial_periodicity"] = "";
+    }
+	metadata["SLiM"]["separate_sexes"] = sex_enabled_ ? true : false;
+	metadata["SLiM"]["nucleotide_based"] = nucleotide_based_ ? true : false;
+    std::string new_metadata_str = metadata.dump(4);
+
+    ret = tsk_table_collection_set_metadata(
+            p_tables, new_metadata_str.c_str(), (tsk_size_t)new_metadata_str.length());
+    if (ret != 0)
+        handle_error("tsk_table_collection_set_metadata", ret);
+
+    // As above, we maybe ought to edit the metadata schema adding our keys,
+    // but then comparing tables is a headache; see tskit#763
+    ret = tsk_table_collection_set_metadata_schema(
+            p_tables, SLIM_TSK_METADATA_SCHEMA.c_str(), (tsk_size_t)SLIM_TSK_METADATA_SCHEMA.length());
+    if (ret != 0)
+        handle_error("tsk_table_collection_set_metadata_schema", ret);
+
+    ////////////
+    // Set metadata schema on each table
+    ret = tsk_edge_table_set_metadata_schema(&p_tables->edges,
+            SLIM_TSK_EDGE_METADATA_SCHEMA.c_str(),
+            SLIM_TSK_EDGE_METADATA_SCHEMA.length());
```

Comment:
So, length() returns a size_t but `tsk_edge_table_set_metadata_schema()` takes a `tsk_size_t` which is smaller, so the compiler is complaining.  Please add a cast here, `(tsk_size_t)SLIM_TSK_EDGE_METADATA_SCHEMA.length()`.  Ditto for the other calls to `tsk_edge_table_set_metadata_schema()` below; six occurrences total, I think.